Add support for eLeReS RX as another RX_SPI receiver

marbalon opened the pull request:
@martinbudden @digitalentity  - waiting for your suggestions ;-)

marbalon committed:
Add support for eLeReS RX as another RX_SPI receiver

marbalon committed:
Some minor tweeks in names etc.

marbalon committed:
Changed define ELERES_RX to USE_RX_ELERES like for others, and removed non standard RX_SPI_IDLE_LOOP_REQUIRED

martinbudden commented on src/main/fc/cli.c:
```diff
@@ -2702,6 +2718,41 @@ static void cliDfu(char *cmdline)
+		}
+		cliPrint("\r\n");
+		cliPrint("Bind OK!\r\nPlease restart your transmitter.\r\n");
+	}
```
Please use K&R brace style, eg:
```
    if (eLeReS_Bind()) {
        cliPrint("Bind timeout!\r\n");
    } else {
        cliPrint("Signature: ");
        for (int i=0; i<4; i++) {
            itoa(eleresConfigMutable()->eleres_signature[i], buf, 16);
            cliPrint(buf);
            cliPrint(" ");
        }
        cliPrint("\r\n");
        cliPrint("Bind OK!\r\nPlease restart your transmitter.\r\n");
    }
```

Note also `i` is declared in `for` loop.

martinbudden commented on src/main/rx/rx.c:
```diff
@@ -515,6 +516,14 @@ void parseRcChannels(const char *input)
+{
+    UNUSED(currentTime);
+    rssi = (eleres_rssi() - 18)*1024/106; //RSSI is in range 18-124
+}
```
Scaling of RSSI should be done within `eleres_rssi()` function.

martinbudden commented on src/main/rx/eleres.c:
```diff
@@ -0,0 +1,854 @@
+void rx_reset(void);
+void to_tx_mode(uint8_t bytes_to_send);
+void to_ready_mode(void);
+void frequency_configurator(uint32_t frequency);
```
We prefer to avoid forward declaration of functions, instead change order of function definitions so that forward declarations are not necessary.

martinbudden commented on src/main/rx/eleres.c:
```diff
@@ -0,0 +1,854 @@
+#define Transmitted	2
+#define Receive 	4
+#define Received 	8
+#define Preamble	16
```
Style is to use CAPITALS for `#defines`.

martinbudden commented on src/main/rx/eleres.c:
```diff
@@ -0,0 +1,854 @@
+uint8_t tx_full = 0;
+
+
+uint8_t rfm_spi_read(uint8_t address)
```
Local functions should be declared `static`. Also general iNav style is to use `camelCase` (it may not look like that because there is quite a bit of historical code, but `camelCase` is preferred for new code).

martinbudden commented on src/main/rx/eleres.c:
```diff
@@ -0,0 +1,854 @@
+void Bind_Channels(const uint8_t* RF_HEAD, uint8_t* hop_lst)
+{
+    uint8_t n,i,j;
+    for (i=0; i<16; i++) hop_lst[i] = 0;
```
Use `memset`.

martinbudden commented on src/main/rx/eleres.c:
```diff
@@ -0,0 +1,854 @@
+    uint8_t St1,St2;
+    rx_spi_received_e res = RX_SPI_RECEIVED_NONE;
+
+    St1 = rfm_spi_read(0x03);
```
Declare variables at point of first usage and use `const` if possible. ie:
```
const uint8_t St1 = rfm_spi_read(0x03);
```

martinbudden commented on src/main/rx/eleres.c:
```diff
@@ -0,0 +1,854 @@
+#include "config/config_master.h"
+#include "config/parameter_group.h"
+#include "config/parameter_group_ids.h"
+#include "scheduler/scheduler.h"
```
I think the vast majority of these header files are not required.

martinbudden commented on src/main/rx/eleres.c:
```diff
@@ -0,0 +1,854 @@
+uint8_t BkgLoc_enable = 0;
+uint8_t BkgLoc_chlist;
+uint8_t BkgLoc_buf[3][10];
+uint8_t BkgLoc_cnt;
```
I think all these variables can be declared `static`.

martinbudden commented on src/main/rx/eleres.c:
```diff
@@ -0,0 +1,854 @@
+        memcpy(RF_Tx_Buffer+3,&debug[1],2);
+        memcpy(RF_Tx_Buffer+5,&debug[2],2);
+        memcpy(RF_Tx_Buffer+7,&debug[3],2);
+        telem_state = 0;
```
Why not `memcpy(RF_Tx_Buffer + 1, &debug[0], 8);`?

marbalon committed:
Code refactoring

marbalon committed:
Version with task. static added for functions

krzysztofmatula commented on src/main/fc/cli.c:
```diff
@@ -890,6 +892,20 @@ static const clivalue_t valueTable[] = {
+    { "eleres_signature_1",         VAR_UINT8  | MASTER_VALUE, .config.minmax = { 0, 255 }, PG_ELERES_CONFIG, offsetof(eleresConfig_t, eleresSignature[0]) },
+    { "eleres_signature_2",         VAR_UINT8  | MASTER_VALUE, .config.minmax = { 0, 255 }, PG_ELERES_CONFIG, offsetof(eleresConfig_t, eleresSignature[1]) },
+    { "eleres_signature_3",         VAR_UINT8  | MASTER_VALUE, .config.minmax = { 0, 255 }, PG_ELERES_CONFIG, offsetof(eleresConfig_t, eleresSignature[2]) },
+    { "eleres_signature_4",         VAR_UINT8  | MASTER_VALUE, .config.minmax = { 0, 255 }, PG_ELERES_CONFIG, offsetof(eleresConfig_t, eleresSignature[3]) },
```
Why not using single 32-bit "eleres_signature" parameter ?

marbalon replied:
Because signature is written to registers byte by byte, and simple trick like this is not working (probably because of PG_xxxx macros etc.):

`(uint8_t*) signature = &eleresConfigMutable()->eleresSignature;`
or 
`uint8_t signature[4];`
`memcpy(signature ,&eleresConfigMutable()->eleresSignature, 4);`

Edit.
Sorry I was wrong... should work but need to fix CLI interface for uint32 becuse it have some bugs...

martinbudden commented on src/main/rx/eleres.c:
```diff
@@ -0,0 +1,730 @@
+                hoppingChannel = rfRxBuffer[21] & 0x0F;
+
+            channelHoppingTime = (rfRxBuffer[20] & 0x0F)+18;
+            dataReady |= 1;
```
Can you replace the magic numbers associated with `dataReady` with `#define`s. Ie instead of
`dataReady |= 1;` use `dataReady |= SOME_MEANINGFUL_NAME;`. Likewise for 2, 4 etc.

marbalon committed:
Fixed VAR_UINT32 parsing in CLI
Move eleres signature to uint32
TASK_RX_ELERES removed, and eleresSetRcDataFromPayload moved to eleresDataReceived

marbalon committed:
Definitions for dataReady bits

digitalentity reviewed (approved):
@martinbudden @marbalon I think it's ready for merge.
